File: go/magicseam/barrier.go
// Copyright (C) 2025-2026 Malformed C. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

package magicseam

import (
	"errors"
	"fmt"
	"sync/atomic"
	"time"
)

// COORDINATED-CHECKPOINT MARKERS FOR A NON-NATIVE PROVIDER (ADR-0032).
//
// # Why a provider written in Go needs this at all
//
// Before markers, a provider could be quiesced only by ITS OWN trail process
// receiving SIGUSR1. A provider built on this SDK has no trail: it is an
// ordinary Go program with no signal handler and no barrier state, so nothing
// could arm it. A coordinated checkpoint of a graph containing one would drain
// the CONSUMER's in-flight calls - the consumer is trail, so it quiesces - while
// this provider served straight through the snapshot instant. A torn cut, and
// silent, because nothing in the system could observe that one member had not
// stopped.
//
// A marker is how a consumer arms a peer it does not run. Implementing the three
// ops is what makes a provider in ANY language a first-class member of a barrier
// rather than the reason one cannot be taken.
//
// # The contract, and the one part that is easy to get wrong
//
// THE ACK MEANS "MY CHANNEL IS EMPTY", NOT "I RECEIVED YOUR MARKER". It is sent
// only once in-flight calls reach zero. A consumer's own in-flight count says
// nothing about what this provider is still processing, which is the entire
// reason the marker exists - so acking on receipt would hand the coordinator a
// false quiesce and produce exactly the torn cut the protocol prevents.
//
// A provider that CANNOT drain must fail the marker rather than ack late or
// hang: the coordinator then aborts and resumes the graph, instead of every
// member sitting armed behind one stuck call.

// Barrier is a provider's quiesce state for the marker protocol.
//
// The zero value is ready to use and means "not armed, nothing in flight".
type Barrier struct {
	// Lease bounds how long an arm survives without a Resume. Zero means
	// DefaultLeaseTimeout. See barrierlease.go for what happens without one -
	// it is not a hypothetical.
	Lease time.Duration

	armed    atomic.Bool
	inFlight atomic.Int64
	// armedAt is the arm instant in unix nanos, 0 when not armed. Read on every
	// Armed() call, which is the hot path of a busy provider, so it is an atomic
	// rather than a mutex-guarded time.Time.
	armedAt atomic.Int64
}

// DefaultDrainTimeout bounds how long a provider waits for its own in-flight
// calls to drain before failing a marker. Matches trail's own bound
// (cmd/trail/src/marker.rs) so both ends of a barrier give up at the same
// point rather than one waiting on the other's already-abandoned attempt.
const DefaultDrainTimeout = 10 * time.Second

// drainPoll is how often the drain re-checks. Short, because a whole barrier is
// waiting on it.
const drainPoll = 5 * time.Millisecond

// ErrDrainTimeout is returned when in-flight calls did not reach zero in time.
// Distinct from a transport error because the coordinator's response differs: a
// transport failure is a lost peer, this is a peer that is present and busy.
var ErrDrainTimeout = errors.New("magicseam: in-flight calls did not drain")

// Armed reports whether new calls should be refused.
//
// Also where the arm LEASE is enforced: an arm that has outlived its lease is
// released here, by whoever asks. See barrierlease.go - a provider armed by a
// consumer that then died has no other way back, because the Resume can only
// arrive on that consumer's own connection.
func (b *Barrier) Armed() bool {
	if !b.armed.Load() {
		return false
	}
	if b.leaseExpired() {
		return false
	}

	return true
}

// InFlight reports how many calls are currently executing.
func (b *Barrier) InFlight() int64 { return b.inFlight.Load() }

// Enter marks a call as started and returns its completion function.
//
// Deferred by the serve path for the call's WHOLE lifetime: a marker that
// arrives while a call is mid-flight must not see zero, or it would ack a
// channel that still has work in it.
func (b *Barrier) Enter() func() {
	b.inFlight.Add(1)

	return func() { b.inFlight.Add(-1) }
}

// Arm stops admitting new calls and waits for in-flight ones to drain.
//
// Returns ErrDrainTimeout if they do not, and leaves the barrier ARMED when it
// does: the caller must not ack, and the coordinator's abort path will send a
// Resume. Un-arming here instead would silently resume a provider the
// coordinator still believes it is negotiating with.
func (b *Barrier) Arm(timeout time.Duration) error {
	b.armed.Store(true)
	// Stamped on EVERY arm, including a re-arm of an already-armed barrier: a
	// coordinator that re-arms is a coordinator that is still alive, and the
	// lease exists only to catch one that is not.
	b.armedAt.Store(time.Now().UnixNano())
	deadline := time.Now().Add(timeout)
	for {
		if b.inFlight.Load() == 0 {
			return nil
		}
		if time.Now().After(deadline) {
			return fmt.Errorf("%w: %d still running after %s - this channel cannot be quiesced, "+
				"so the coordinator must abort rather than snapshot a torn cut",
				ErrDrainTimeout, b.inFlight.Load(), timeout)
		}
		time.Sleep(drainPoll)
	}
}

// Resume releases the barrier. Never waits to drain: resuming is not a
// quiescence claim, and a coordinator on its abort path must be able to release
// a provider that is busy - that is precisely the case it is aborting for.
func (b *Barrier) Resume() {
	b.armed.Store(false)
	b.armedAt.Store(0)
}

File: ts/magicseam/barrier.ts
// Copyright (C) 2025-2026 Malformed C. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

// COORDINATED-CHECKPOINT MARKERS for a provider written in TypeScript
// (ADR-0032, docs/magic-seam-quic-protocol.md §8).
//
// A marker is how a consumer quiesces a provider it does not run. Without it
// this SDK's providers cannot join a coordinated checkpoint at all: the consumer
// drains its own in-flight calls while this provider keeps serving through the
// snapshot instant, producing a torn cut that nothing detects.
//
// go/magicseam/barrier.go is the reference implementation and
// trail's marker handling is the consumer side; all three are implementations
// of one written protocol and must agree on the wire, not on their internals.
//
// # The contract, and the part that is easy to get wrong
//
// THE ACK MEANS "MY CHANNEL IS EMPTY", NOT "I RECEIVED YOUR MARKER". It is sent
// only once in-flight calls reach zero. A consumer's own in-flight count says
// nothing about what this provider is still processing - which is the entire
// reason the marker exists - so acking on receipt hands the coordinator a false
// quiesce and the torn cut the protocol prevents.

/** How long a drain may take before the marker FAILS (§8 rule 2). Matches
 *  go/magicseam's DefaultDrainTimeout and trail's own bound, so both ends of
 *  a barrier give up at the same point rather than one waiting on the other's
 *  already-abandoned attempt. */
export const DEFAULT_DRAIN_TIMEOUT_MS = 10_000

/** How long an arm survives with no resume.
 *
 *  MUST stay below the Go SDK's DefaultLeaseTimeout (2m) and above trail's
 *  consumer lease (90s)? No - it must simply bound an ABANDONED arm. The
 *  consumer gives up first at 90s and sends a Resume; this is the fallback for a
 *  consumer that is GONE and can never send one, exactly as
 *  go/magicseam/barrierlease.go describes. Measured there: without a lease a
 *  provider armed by a consumer that died refused every caller forever, cleared
 *  only by a restart. */
export const DEFAULT_LEASE_MS = 120_000

/** Returned by arm() when in-flight calls did not reach zero in time. Distinct
 *  from a transport failure because the coordinator's response differs: a lost
 *  peer versus a peer that is present and busy. */
export class DrainTimeoutError extends Error {
  constructor(inFlight: number, timeoutMs: number) {
    super(
      `magicseam: ${inFlight} in-flight call(s) did not drain in ${timeoutMs}ms - this channel ` +
        `cannot be quiesced, so the coordinator must abort rather than snapshot a torn cut`,
    )
    this.name = "DrainTimeoutError"
  }
}

/**
 * A provider's quiesce state. The zero state ("not armed, nothing in flight") is
 * what a fresh instance has, so a provider that never takes a barrier behaves
 * exactly as it did before markers existed.
 */
export class Barrier {
  #armed = false
  #inFlight = 0
  #armedAt = 0
  readonly #leaseMs: number

  constructor(leaseMs: number = DEFAULT_LEASE_MS) {
    this.#leaseMs = leaseMs
  }

  /**
   * Whether new calls should be refused.
   *
   * Also where the arm LEASE is enforced, at READ time rather than by a timer:
   * nothing to leak, nothing to cancel, and no path where the releasing
   * mechanism is itself wedged. An arm that has outlived its lease was abandoned
   * by a consumer that can no longer send a Resume - the Resume travels on that
   * consumer's own connection and no other party knows the barrier exists.
   */
  get armed(): boolean {
    if (!this.#armed) return false
    if (this.#armedAt !== 0 && Date.now() - this.#armedAt > this.#leaseMs) {
      const heldMs = Date.now() - this.#armedAt
      this.#armed = false
      this.#armedAt = 0
      console.error(
        `[magicseam][barrier] LEASE EXPIRED: armed ${Math.round(heldMs / 1000)}s ago with no ` +
          `Resume - releasing and admitting calls again. The coordinator that armed this ` +
          `provider is gone; without this the provider would refuse every caller forever.`,
      )

      return false
    }

    return true
  }

  /** How long this barrier has been armed, or 0 if it is not.
   *
   *  A SECOND, INDEPENDENT READER: it never runs the lease release, so a lease
   *  that silently fails to fire is still visible. Readiness checks must use
   *  this rather than `armed`, or the lease becomes its own alibi - released
   *  nothing, reports not-armed, health green, provider serving nobody. */
  get armedForMs(): number {
    return this.#armedAt === 0 ? 0 : Date.now() - this.#armedAt
  }

  get inFlight(): number {
    return this.#inFlight
  }

  /** Mark a call as started; the returned function marks it finished. Held for
   *  the call's WHOLE lifetime, or a marker arriving mid-call sees zero and acks
   *  a channel that still has work in it. */
  enter(): () => void {
    this.#inFlight++
    let done = false

    return () => {
      if (done) return
      done = true
      this.#inFlight--
    }
  }

  /**
   * Stop admitting new calls and wait for in-flight ones to drain.
   *
   * Throws DrainTimeoutError if they do not, and leaves the barrier ARMED when
   * it does (§8 rule 3): the caller must not ack, and the coordinator's abort
   * path sends the Resume. Un-arming here would silently resume a provider the
   * coordinator still believes it is negotiating with.
   */
  async arm(timeoutMs: number = DEFAULT_DRAIN_TIMEOUT_MS): Promise<void> {
    this.#armed = true
    // Stamped on EVERY arm, including a re-arm: a coordinator that re-arms is
    // one that is still alive, and the lease exists only to catch one that is not.
    this.#armedAt = Date.now()

    const deadline = Date.now() + timeoutMs
    for (;;) {
      if (this.#inFlight === 0) return
      if (Date.now() > deadline) throw new DrainTimeoutError(this.#inFlight, timeoutMs)
      await new Promise((r) => setTimeout(r, 5))
    }
  }

  /** Release the barrier. NEVER waits to drain: resuming is not a quiescence
   *  claim, and the abort path exists precisely for a provider that is busy. */
  resume(): void {
    this.#armed = false
    this.#armedAt = 0
  }
}
